packages/datastore/src/sync/utils.ts

```diff
@@ -522,6 +579,211 @@ export function predicateToGraphQLFilter(
+		const current = stack.pop();
+
+		if (isPredicateGroup(current)) {
+			const { predicates, type } = current;
+			// ignore length = 1; groups with 1 predicate will get flattened when converted to gqlFilter
+			if (type === 'or' && predicates.length > 1) {
+				count += predicates.length;
+			}
+			stack.push(...predicates);
+		}
+	}
+
+	// if we didn't encounter any OR groups, default to 1
+	return count || 1;
+}
+
+/**
+ *
+ * @param group - selective sync predicate group
+ * @returns name of repeated field | null
+ *
+ * @example returns "username"
+ * ```js
+ * { type: "and", predicates: [
+ * 		{ field: "username", operator: "beginsWith", operand: "a" },
+ * 		{ field: "username", operator: "contains", operand: "abc" },
+ * ] }
+ * ```
+ */
+export function repeatedFieldInGroup(
+	group?: PredicatesGroup<any>
+): string | null {
+	if (!group || !Array.isArray(group.predicates)) return null;
+
+	// convert to filter in order to flatten redundant groups
+	const gqlFilter = predicateToGraphQLFilter(group);
+
+	const stack: GraphQLFilter[] = [gqlFilter];
+
+	const hasGroupRepeatedFields = (fields: GraphQLFilter[]): string | null => {
+		const seen = {};
+
+		for (const f of fields) {
+			const [fieldName] = Object.keys(f);
+			if (seen[fieldName]) {
+				return fieldName;
+			}
+			seen[fieldName] = true;
+		}
+		return null;
+	};
+
+	while (stack.length > 0) {
+		const current = stack.pop();
+
+		const [key] = Object.keys(current!);
+		const values = current![key];
+
+		if (!Array.isArray(values)) {
+			return null;
+		}
+
+		// field value will be single object
+		const predicateObjects = values.filter(
+			v => !Array.isArray(Object.values(v)[0])
+		);
+
+		// group value will be an array
+		const predicateGroups = values.filter(v =>
+			Array.isArray(Object.values(v)[0])
+		);
+
+		if (key === 'and') {
+			const repeatedField = hasGroupRepeatedFields(predicateObjects);
+			if (repeatedField) {
+				return repeatedField;
+			}
+		}
+
+		stack.push(...predicateGroups);
+	}
+
+	return null;
+}
+
+export enum RTFError {
+	UnknownField,
+	MaxAttributes,
+	MaxCombinations,
+	RepeatedFieldname,
+	NotGroup,
+}
+
+export function generateRTFRemediation(
+	errorType: RTFError,
+	modelDefinition: SchemaModel,
+	predicatesGroup: PredicatesGroup<any> | undefined
+): string {
+	const selSyncFields = filterFields(predicatesGroup);
+	const selSyncFieldStr = [...selSyncFields].join(', ');
+	const dynamicAuthModeFields = dynamicAuthFields(modelDefinition);
+	const dynamicAuthFieldsStr = [...dynamicAuthModeFields].join(', ');
+	const filterCombinations = countFilterCombinations(predicatesGroup);
+	const repeatedField = repeatedFieldInGroup(predicatesGroup);
+
+	switch (errorType) {
+		case RTFError.UnknownField:
+			return (
+				`Your API was generated with an older version of the CLI that doesn't support backend subscription filtering.` +
+				'To enable backend subscription filtering: upgrade your Amplify CLI to the latest version and run `amplify push`'
```

Comment:
Might be good to give them both commands they need to run.

```suggestion
				'To enable backend subscription filtering, upgrade your Amplify CLI to the latest version and push your app by running `amplify upgrade` followed by `amplify push`'.
```